PhpAmqpLib/Wire/IO/SocketIO.php

```diff
@@ -194,4 +219,40 @@ protected function enable_keepalive()
 
         socket_set_option($this->sock, SOL_SOCKET, SO_KEEPALIVE, 1);
     }
+
+    /**
+     * Heartbeat logic: check connection health here
+     */
+    protected function check_heartbeat()
+    {
+        // ignore unless heartbeat interval is set
+        if ($this->heartbeat !== 0 && $this->last_read && $this->last_write) {
+            $t = microtime(true);
+            $t_read = round($t - $this->last_read);
+            $t_write = round($t - $this->last_write);
+
+            // server has gone away
+            if (($this->heartbeat * 2) < $t_read) {
+                $this->reconnect();
+            }
+
+            // time for client to send a heartbeat
+            if (($this->heartbeat / 2) < $t_write) {
+                $this->write_heartbeat();
+            }
+        }
+    }
+
+    /**
+     * Sends a heartbeat message
+     */
+    protected function write_heartbeat()
+    {
+        $pkt = new AMQPWriter();
```

Comment:
`AMQPWriter` is a stateful object and there's no way to reset it. 

This part of the code was copied from `StreamIO`. This is not the best apologize but I don't see how else to write a packet. Unless doing it manually which is IMHO worse than "give more work to the GC".

What do you think?